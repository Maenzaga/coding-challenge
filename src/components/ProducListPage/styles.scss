$primary-clr: #28965a;
$secondary-clr: #faf8d4;
$my-font: "Nunito", sans-serif;

.main-content {
  display: flex;
  flex-direction: column;
  padding: 6px;
  background-color: $primary-clr;
  &__search {
    align-self: center;
    &-box {
      border-radius: 5%;
      background-color: hsla(57, 96%, 54%, 0.925);
      margin-top: 2rem;
      margin-bottom: 2rem;
      color: #08ae53;
      font-family: $my-font;
      font-weight: bold;
      font-size: 1.5rem;
      height: 2rem;
      width: 16rem;
      border: 0;
      &::placeholder {
        color: #08ae53;
      }
    }
  }
  &__list {
    background-color: $primary-clr;
    padding: 10px;
    display: grid;
    gap: 3px;
    grid-template-columns: repeat(5, 1fr);
  }
}
