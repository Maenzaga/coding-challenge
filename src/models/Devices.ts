export type DeviceApi = {
    "id": string;
    "brand":string;
    "model":string;
    "price": string;
    "imgUrl": string
}

export type Device = {
    id: string;
    brand:string;
    model:string;
    price: string;
    imgUrl: string
}

export type DeviceDetailApi = {
    "id": string;
    "brand":string;
    "model":string;
    "price": string;
    "imgUrl": string;
    "cpu": string;
    "ram":string;
    "os":string;
    "displayResolution": string;
    "battery": string;
    "primaryCamera": string[];
    "secondaryCamera": string[];
    "dimentiones":string;
    "weight":string;
}
export type DeviceDetail = {
    id: string;
    brand:string;
    model:string;
    price: string;
    imgUrl: string;
    cpu: string;
    ram:string;
    os:string;
    displayResolution: string;
    battery: string;
    primaryCamera: string[];
    secondaryCamera: string[];
    dimentiones:string;
    weight:string;
}

export type DevicesApiResponse = DeviceApi[]
export type DeviceDetailResponse = DeviceDetailApi