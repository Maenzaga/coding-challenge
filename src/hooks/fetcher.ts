import { AxiosResponse } from "axios";
import api from "../api/devices";
import { DevicesApiResponse } from "../models/Devices";

export const fetcher = async (url: string) => {
  const response: AxiosResponse<DevicesApiResponse> = await api.get(url);
  return response.data;
};
