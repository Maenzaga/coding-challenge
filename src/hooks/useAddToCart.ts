import { useCallback, useContext } from "react";
import { CartContext } from "../contexts/CartContext";
import { CartRequestBody } from "../models/Devices";
import api from "../api/devices";

export const useAddToCart = (props: CartRequestBody) => {
  const [count, setCartCount] = useContext(CartContext);

  const addToCart = useCallback(async () => {
    try {
      await api.post("/cart", props);
      setCartCount(count + 1);
    } catch (error) {}
  }, []);

  return addToCart;
};
