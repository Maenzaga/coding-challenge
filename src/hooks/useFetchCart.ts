import { CartItemApi, CartRequestBody, DevicesApiResponse } from "../models/Devices";
import api from "../api/devices"
import { AxiosResponse } from "axios";
import { useEffect, useState } from "react";

export const useFetchCart =({id, colorCode, storageCode}: CartRequestBody) => {
    const [cartCount, setCartCount] = useState(0)
    /* const [error, setError] = useState(''); */

    const cartRequestObj = {id, colorCode, storageCode};

    const fetchCart = async () =>{
        const response: AxiosResponse<CartItemApi> = await api.post(
            "/cart",
            cartRequestObj
            )
            /* .catch(error=>{
                setError({error.message})
            }); */
            setCartCount(response.data.count);
            console.log(response.data.count);
        };

        useEffect(()=>{
            fetchCart()
        },[])

        return cartCount
}