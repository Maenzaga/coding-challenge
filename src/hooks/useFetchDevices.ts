import useSWR from "swr";
import { DevicesApiResponse } from "../models/Devices";

export const useFetchDevices = () => {
  const { data, error } = useSWR<DevicesApiResponse>(`/product`);

  return { data, isError: error, isLoading: !error && !data };
};
