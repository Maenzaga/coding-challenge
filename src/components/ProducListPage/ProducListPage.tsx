import { useState } from "react";
import { useFetchDevices } from "../../hooks/useFetchDevices";
import { ProductCard } from "../ProductCard/ProductCard";
import { ProductFooter } from "../ProductFooter/ProductFooter";
import { ProductHeader } from "../ProductHeader/ProductHeader";
import "./styles.scss";

export const ProductListPage = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const { data, isLoading, isError } = useFetchDevices();

  return (
    <div className="app-container">
      <ProductHeader />
      {!isLoading ? (
        <main className="main-content">
          <div className="main-content__search">
            <input
              type="text"
              placeholder="Search..."
              className="main-content__search-box"
              onChange={(event) => {
                setSearchTerm(event.target.value);
              }}
            />
          </div>
          <div className="main-content__list">
            {data
              ?.filter((val) => {
                if (searchTerm == "") {
                  return val;
                } else if (
                  val.model.toLowerCase().includes(searchTerm.toLowerCase()) ||
                  val.brand.toLowerCase().includes(searchTerm.toLowerCase())
                ) {
                  return val;
                }
              })
              .map((device) => {
                return (
                  <ProductCard
                    key={device.id}
                    id={device.id}
                    imgUrl={device.imgUrl}
                    brand={device.brand}
                    model={device.model}
                    price={device.price}
                  />
                );
              })}
          </div>
          {isError && <h1>{isError}</h1>}
        </main>
      ) : (
        <main className="main-content__loading">Loading...</main>
      )}
      <ProductFooter />
    </div>
  );
};
