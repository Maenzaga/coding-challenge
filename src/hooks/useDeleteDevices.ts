import api from "../api/devices"
import { useEffect, useState } from "react"
import { Device, DeviceDetailApi, DeviceDetailResponse, DevicesApiResponse } from "../models/Devices"
import { useFetchDevices } from "./useFetchDevices"

export const useDeleteDevices = (id:string) =>{
    const [devicesNew, setDevices] = useState<Device[]>([])
    const {data} = useFetchDevices();

    const removeDevice = async (id: string) =>{
        await api.delete(`/product/${id}`);
        const newDevicesList = devicesNew.filter((device)=>{
            return device.id !== id;
        });
        setDevices(newDevicesList);
    }

    useEffect(()=>{
        if (data){
            setDevices(data)
        }
    }, [data])

    return {removeDevice}
}