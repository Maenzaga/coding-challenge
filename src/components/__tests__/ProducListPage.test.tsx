import { render } from "@testing-library/react";
import { useFetchDevices } from "../../hooks/useFetchDevices";
import { Device } from "../../models/Devices";
import { ProductListPage } from "../ProducListPage/ProducListPage";

jest.mock("../../hooks/useFetchDevices", () => ({
  useFetchDevices: jest.fn(),
}));

const mockedUseFetchDevices = useFetchDevices as jest.Mock;

// To update snapshots, run: npm test -- -u

describe("ProductListPage", () => {
  it("renders correctly", () => {
    // Given
    const devices: Device[] = [
      {
        id: "123",
        brand: "Apple",
        model: "iPhone 12",
        price: "1000",
        imgUrl: "some_url_image",
      },
    ];
    mockedUseFetchDevices.mockReturnValue({
      data: devices,
      isError: false,
      isLoading: false,
    });

    // When
    const tree = render(<ProductListPage />);

    // Then
    expect(tree).toMatchSnapshot();
  });

  it("renders loading view when request is in progress", () => {
    // Given
    mockedUseFetchDevices.mockReturnValue({
      data: [],
      isError: false,
      isLoading: true,
    });

    // When
    const { getByText } = render(<ProductListPage />);

    // Then
    expect(getByText("Loading...")).toBeDefined();
  });
});
