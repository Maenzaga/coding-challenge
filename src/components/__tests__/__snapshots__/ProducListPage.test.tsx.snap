// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListPage renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app-container"
      >
        <header
          class="header-container"
        >
          <div
            class="scene"
          >
            <div
              class="screen"
            />
            <div>
              0
            </div>
          </div>
        </header>
        <main
          class="main-content"
        >
          <div
            class="main-content__search"
          >
            <input
              class="main-content__search-box"
              placeholder="Search..."
              type="text"
            />
          </div>
          <div
            class="main-content__list"
          >
            <div
              class="card-container"
            >
              <img
                class="card-image"
                src="some_url_image"
              />
              <section
                class="card-info"
              >
                <p
                  class="card-info__model"
                >
                  iPhone 12
                </p>
                <p
                  class="card-info__brand"
                >
                  Apple
                </p>
                <p
                  class="card-info__price"
                >
                  1000
                   $
                </p>
              </section>
            </div>
          </div>
        </main>
        <footer
          class="product-footer"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app-container"
    >
      <header
        class="header-container"
      >
        <div
          class="scene"
        >
          <div
            class="screen"
          />
          <div>
            0
          </div>
        </div>
      </header>
      <main
        class="main-content"
      >
        <div
          class="main-content__search"
        >
          <input
            class="main-content__search-box"
            placeholder="Search..."
            type="text"
          />
        </div>
        <div
          class="main-content__list"
        >
          <div
            class="card-container"
          >
            <img
              class="card-image"
              src="some_url_image"
            />
            <section
              class="card-info"
            >
              <p
                class="card-info__model"
              >
                iPhone 12
              </p>
              <p
                class="card-info__brand"
              >
                Apple
              </p>
              <p
                class="card-info__price"
              >
                1000
                 $
              </p>
            </section>
          </div>
        </div>
      </main>
      <footer
        class="product-footer"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
